### PR TITLE
reimplement top_contributors_questions in ES7

### PR DESCRIPTION
**Pull request by LeoMcA**
https://github.com/mozilla/sumo-project/issues/652

### CODE REVIEW

File: kitsune/community/utils.py [hunk 1]
```diff
@@ -4,56 +4,86 @@
...
+            {
+                "count": bucket.doc_count,
+                "term": bucket.key,
+                "user": {
+                    "id": user["id"],
+                    "username": user["username"],
+                    "display_name": user["profile__name"],
+                    "avatar": user["profile__fxa_avatar"],
+                    "days_since_last_activity": (
+                        datetime.now()
```
Comment (akatsoulas) [inline kitsune/community/utils.py:77]: nit: maybe we calculate this outside of the return statement. My only reason for proposing this is code readability

File: kitsune/community/utils.py [hunk 2]
```diff
@@ -4,56 +4,86 @@
...
+                        - datetime.fromisoformat(bucket.latest.hits.hits[0]._source.created)
+                    ).days,
+                },
+            }
+            for bucket in contribution_buckets[:count]
```
Comment (akatsoulas) [inline kitsune/community/utils.py:82]: let's move all of this outside of the return. It's a bit challenging wrapping your head around this block of code.

**Review by akatsoulas — COMMENTED** [review]
just a couple of comments for now. I haven't gone through functionality yet

File: kitsune/community/utils.py (hunk 2 above)
Comment (LeoMcA) [reply kitsune/community/utils.py]: I was feeling very functional that day - refactored to use a bog standard for loop instead

File: kitsune/community/utils.py [hunk 3]
```diff
@@ -4,56 +4,87 @@
...
-    return counts
+    contribution_buckets = search.execute().aggregations.filters.contributions.buckets
+    total_contributors = len(contribution_buckets)
+
+    user_ids = map(lambda x: int(x.key), contribution_buckets[:count])
+    users = {
+        # fetch the users found in the ES aggregation,
+        # and put them in a dict with their id as the key
+        user["id"]: user
+        for user in User.objects.filter(id__in=user_ids).values(
```
Comment (akatsoulas) [inline kitsune/community/utils.py:61]: can't we get the data directly from ES?

**Review by akatsoulas — APPROVED** [review]
r+wc 
Personally I would like the idea to go fully with ES on this instead of hitting the DB too unless there are reasons not too. I would be surprised if ES is slower on this although I don't expect big differences. To the extend that we can we can reduce queries towards the DB leveraging ES. Up to you though

File: kitsune/community/utils.py [hunk 4]
```diff
@@ -4,56 +4,87 @@
...
-            product = product.slug
-        answers = answers.filter(question__product__slug=product)
-
-    users = (
-        User.objects.filter(answers__in=answers)
-        .annotate(query_count=Count("answers"))
-        .order_by("-query_count")
+        filters.append({"term": {"question_product_id": product.id}})
+
+    search.aggs.bucket(
```
[the remark this answers is not in the document]
Comment (LeoMcA) [reply kitsune/community/utils.py:35]: they seem to be as performant as each other, but I think using a filtered search makes the code clearer

**Comment by LeoMcA** [comment]
Actually fetching profiles from ES works rather nicely because it returns them in the order the ids were specified, so I don't need to do that list to dict conversion I was doing.

File: kitsune/community/utils.py [hunk 5]
```diff
@@ -4,56 +4,82 @@
...
-        # By default we go back 90 days.
-        start = date.today() - timedelta(days=90)
-        answers = answers.filter(created__gte=start)
-    if end:
-        # If no end is specified, we don't need to filter by it.
-        answers = answers.filter(created__lt=end)
+    # filter out answers by the question author
+    search = search.filter(
+        "script", script="doc['creator_id'].value != doc['question_creator_id'].value"
+    )
```
Comment (akatsoulas) [inline kitsune/community/utils.py:25]: How about chaining them in order to be more similar to Django's syntax.

File: kitsune/community/utils.py [hunk 6]
```diff
@@ -4,56 +4,82 @@
...
-    if use_cache:
-        cache.set(cache_key, counts, 60 * 180)  # 3 hours
-    return counts
+    contribution_buckets = search.execute().aggregations.contributions.buckets
+    total_contributors = len(contribution_buckets)
+
+    if not total_contributors:
+        return [], 0
+
+    user_ids = map(lambda x: int(x.key), contribution_buckets[:count])
```
Comment (akatsoulas) [inline kitsune/community/utils.py:56]: list comprehension maybe?

File: kitsune/community/utils.py [hunk 7]
```diff
@@ -4,56 +4,82 @@
...
+        if bucket.key != user.meta.id:
+            raise RuntimeError("ES returned profiles out of order")
+        last_activity = datetime.fromisoformat(bucket.latest.hits.hits[0]._source.created)
+        days_since_last_activity = (datetime.now() - last_activity).days
+        top_contributors.append(
+            {
+                "count": bucket.doc_count,
+                "term": bucket.key,
+                "user": {
+                    "id": bucket.key,
```
Comment (akatsoulas) [inline kitsune/community/utils.py:73]: maybe use `user` here too?

**Review by akatsoulas — APPROVED** [review]
r+wc

**Pull request merged**